cmd/server/main.go

```diff
@@ -0,0 +1,173 @@
+package main
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"encoding/binary"
+	"fmt"
+	"github.com/kr/pty"
+	"golang.org/x/crypto/ssh"
+	"io"
+	"log"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"syscall"
+	"unsafe"
+)
+
+func setWinsize(fd uintptr, w, h uint32) {
+	syscall.Syscall(
+		syscall.SYS_IOCTL,
+		fd,
+		uintptr(syscall.TIOCSWINSZ),
+		uintptr(unsafe.Pointer(
+			&struct{ h, w, x, y uint16 }{uint16(h), uint16(w), 0, 0},
+		)),
+	)
+}
+
+func parseDims(b []byte) (uint32, uint32) {
+	w := binary.BigEndian.Uint32(b)
+	h := binary.BigEndian.Uint32(b[4:])
+	return w, h
+}
+
+func ptyRun(c *exec.Cmd, tty *os.File) (err error) {
+	defer tty.Close()
+	c.Stdout = tty
+	c.Stdin = tty
+	c.Stderr = tty
+	c.SysProcAttr = &syscall.SysProcAttr{
+		Setctty: true,
+		Setsid:  true,
+	}
+	return c.Start()
+}
+
+func publicKeyCallback(
+	conn ssh.ConnMetadata,
+	key ssh.PublicKey,
+) (*ssh.Permissions, error) {
+	return &ssh.Permissions{
+		Extensions: map[string]string{
+			"pubkey": strings.TrimSpace(string(ssh.MarshalAuthorizedKey(key))),
+		},
+	}, nil
+}
+
+func keyboardInteractiveCallback(
+	ssh.ConnMetadata,
+	ssh.KeyboardInteractiveChallenge,
+) (*ssh.Permissions, error) {
+	return nil, nil
+}
+
+func handleConnection(nConn net.Conn, sshConfig *ssh.ServerConfig) {
+	conn, chans, _, _ := ssh.NewServerConn(nConn, sshConfig)
+	go func(chans <-chan ssh.NewChannel) {
+		for newChannel := range chans {
+			if t := newChannel.ChannelType(); t != "session" {
+				newChannel.Reject(
+					ssh.UnknownChannelType,
+					fmt.Sprintf("unknown channel type: %s", t),
+				)
+				continue
+			}
+			channel, requests, _ := newChannel.Accept()
+			f, tty, _ := pty.Open()
+			go func(in <-chan *ssh.Request) {
+				user := conn.Conn.User()
+				ip := fmt.Sprintf("%s", conn.Conn.RemoteAddr())
+				version := fmt.Sprintf("%s", conn.Conn.ClientVersion())
+				key := "none"
+				if conn.Permissions != nil {
+					if conn.Permissions.Extensions != nil {
```

Comment:
These two if statements could be condensed into a one-layer `if conn.Permissions != nil; conn.Permissions.Extensions != nil`